src/torchcodec/decoders/_core/VideoDecoder.cpp

```diff
@@ -979,36 +980,36 @@ VideoDecoder::AVFrameStream VideoDecoder::decodeAVFrame(
     // We still haven't found the frame we're looking for. So let's read more
     // packets and send them to the decoder.
     ReferenceAVPacket packet(autoAVPacket);
-    ffmpegStatus = av_read_frame(formatContext_.get(), packet.get());
-    decodeStats_.numPacketsRead++;
+    bool foundPacketForStream = false;
+    while (!foundPacketForStream) {
+      ffmpegStatus = av_read_frame(formatContext_.get(), packet.get());
+      decodeStats_.numPacketsRead++;
+
+      if (ffmpegStatus == AVERROR_EOF) {
+        // End of file reached. We must drain the codec by sending a nullptr
+        // packet.
+        ffmpegStatus = avcodec_send_packet(
+            streamInfo.codecContext.get(),
+            /*avpkt=*/nullptr);
+        if (ffmpegStatus < AVSUCCESS) {
+          throw std::runtime_error(
+              "Could not flush decoder: " +
+              getFFMPEGErrorStringFromErrorCode(ffmpegStatus));
+        }
+
+        // We've reached the end of file so we can't read any more packets from
+        // it, but the decoder may still have frames to read in its buffer.
+        // Continue iterating to try reading frames.
+        reachedEOF = true;
+        goto outerLoopStart;
+      }
 
-    if (ffmpegStatus == AVERROR_EOF) {
-      // End of file reached. We must drain the codec by sending a nullptr
-      // packet.
-      ffmpegStatus = avcodec_send_packet(
-          streamInfo.codecContext.get(),
-          /*avpkt=*/nullptr);
       if (ffmpegStatus < AVSUCCESS) {
         throw std::runtime_error(
-            "Could not flush decoder: " +
+            "Could not read frame from input file: " +
             getFFMPEGErrorStringFromErrorCode(ffmpegStatus));
       }
-
-      // We've reached the end of file so we can't read any more packets from
-      // it, but the decoder may still have frames to read in its buffer.
-      // Continue iterating to try reading frames.
-      reachedEOF = true;
-      continue;
-    }
-
-    if (ffmpegStatus < AVSUCCESS) {
-      throw std::runtime_error(
-          "Could not read frame from input file: " +
-          getFFMPEGErrorStringFromErrorCode(ffmpegStatus));
-    }
-
-    if (packet->stream_index != activeStreamIndex_) {
-      continue;
```

Comment:
So, in main, we just called `av_read_frame` to receive a packet, and now that we realize this isn't a packet for the target stream, we `continue;` which brings us back to the beginning of the main loop... which starts by calling `avcodec_receive_frame()`.

But there is no point `avcodec_receive_frame()`: it's bound to return `EAGAIN`, because we haven't sent any new packet to the decoder!

So that's what this PR is doing: we don't `continue;` and go back to the loop. Instead, we just call `av_read_frame` again until we get a packet for the target stream.